atc/db/build_in_memory_check.go

```diff
@@ -0,0 +1,313 @@
+		RunWith(tx).
+		Exec()
+	return err
+}
+
+func (b *inMemoryCheckBuild) Variables(logger lager.Logger, secrets creds.Secrets, varSourcePool creds.VarSourcePool) (vars.Variables, error) {
+	pipeline, found, err := b.Pipeline()
+	if err != nil {
+		return nil, fmt.Errorf("failed to find pipeline: %w", err)
+	}
+	if !found {
+		return nil, errors.New("pipeline not found")
+	}
+
+	return pipeline.Variables(logger, secrets, varSourcePool)
+}
+
+func (b *inMemoryCheckBuild) SaveEvent(ev atc.Event) error {
+	tx, err := b.conn.Begin()
+	if err != nil {
+		return err
+	}
+	defer Rollback(tx)
+
+	err = b.saveEvent(tx, ev)
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+	return b.conn.Bus().Notify(buildEventsChannel(b.id))
+}
+
+func (b *inMemoryCheckBuild) Events(from uint) (EventSource, error) {
+	notifier, err := newConditionNotifier(b.conn.Bus(), buildEventsChannel(b.id), func() (bool, error) {
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return newBuildEventSource(
+		b.id,
+		"check_build_events",
+		b.conn,
+		notifier,
+		from,
+		true,
+	), nil
+}
+
+func (b *inMemoryCheckBuild) AbortNotifier() (Notifier, error) {
+	return newNoopNotifier(), nil
+}
+
+func (b *inMemoryCheckBuild) SpanContext() propagation.TextMapCarrier {
+	return SpanContext{}
+}
+
+func (b *inMemoryCheckBuild) IsManuallyTriggered() bool {
+	return false
+}
+
+func (b *inMemoryCheckBuild) PrivatePlan() atc.Plan {
+	return b.plan
+}
+
+func (b *inMemoryCheckBuild) HasPlan() bool            {
+	if b.plan.ID != "" {
+		return true
+	}
+	resource, ok := b.checkable.(Resource)
+	return ok && resource.BuildSummary() != nil && resource.BuildSummary().PublicPlan != nil
+}
+
+func (b *inMemoryCheckBuild) PublicPlan() *json.RawMessage {
+	if b.plan.ID != "" {
+		return b.plan.Public()
+	}
+
+	resource, ok := b.checkable.(Resource)
+	if !ok || resource.BuildSummary() == nil || resource.BuildSummary().PublicPlan == nil {
+		return nil
+	}
+	bytes, err := json.Marshal(resource.BuildSummary().PublicPlan)
+	if err != nil {
+		return nil
+	}
+	m := json.RawMessage(bytes)
+	return &m
+}
+
+func (b *inMemoryCheckBuild) ID() int { return b.id }
+
+func (b *inMemoryCheckBuild) Name() string { return CheckBuildName }
+
+// JobID returns 0 because check build doesn't belong to any job.
+func (b *inMemoryCheckBuild) JobID() int { return 0 }
+
+// JobName returns an empty string because check build doesn't belong to any job.
+func (b *inMemoryCheckBuild) JobName() string { return "" }
+
+func (b *inMemoryCheckBuild) CreateTime() time.Time { return b.createTime }
+
+func (b *inMemoryCheckBuild) IsNewerThanLastCheckOf(input Resource) bool {
+	return b.createTime.After(input.LastCheckEndTime())
+}
+
+// IsRunning returns true as a in-memory check build only runs once.
+func (b *inMemoryCheckBuild) IsRunning() bool {
+	return true
```

Comment:
This is a good point, however I guess `BuildSummary` won't work here. Because `BuildSummary` is from `resource_config_scope`. When a resource start to check, the status should be a `started`, and the status is only updated in `checkStep.Run()`, so that when `engine` call `build.IsRunning()`, there is no chance to update the status.

There are two types of in memory check builds: 

1. Lidar created, that will be run by `engine`
2. `buildFactory.Build()` created, that will be used to query event logs

So I added two `new` functions and added a `running` flag with setting it accordingly in `new` functions.